### PR TITLE
Creates the Joint estimators

### PR DESCRIPTION
**Pull request by victorventuri**
These changes create the joint estimation framework (for general and UKF online estimators), as well as implements the full interface between ECM model and joint UKF estimator. A simple Rint ECM is used as a unit test, and a notebook is provided demonstrating how to use it the ECMJointUKF.

### CODE REVIEW

File: tests/estimators/online/joint/test_ecm_joint.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    def update_timestamps_curr(duration: float, curr: float) -> None:
+        """
+        Helper to populate timestamps and currents.
+
+        Args:
+            duration: duration of segment in seconds
+            curr: value of current to be used.
+        """
+        # specify binds to nearest variables
+        nonlocal currents
```
Comment (WardLT) [inline tests/estimators/online/joint/test_ecm_joint.py:48]: Nice! I didn't know about non-local as an alternative to global.

**Review by WardLT — APPROVED** [review]
I think this is good to merge. I'd like to go through and cut down the number of classes in the next iteration, but I'm happy the way that things work now.

**Comment by WardLT** [comment]
@victorventuri , could you resolve the conflict and the question about adding `validate_assignement` to the base `HealthVariable` class?

File: asoh/estimators/online/joint/ecm/unscented.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            ecm_out = ECM.calculate_terminal_voltage(new_input=ecm_input,
+                                                     transient_state=self.transient,
+                                                     asoh=self.asoh)
+            pred_measurement += [ecm_out.to_numpy()]
+
+        return np.array(pred_measurement)
+
+
+# TODO (vventuri): how do we denoise SOC, Qt, R0?
+class ECMJointUKF(JointUKFEstimator):
```
[the remark this answers is not in the document]
Comment (victorventuri) [reply asoh/estimators/online/joint/ecm/unscented.py:132]: I think that makes sense, and mostly agree with it. We can see how things go in the future, but from the current versions we have, each model will probably still need its own interface with the estimator.

File: asoh/models/ecm/components.py [hunk 3]
```diff
@@ -107,14 +114,14 @@ class ReferenceOCV(SOCInterpolatedHealth):
...
-class EntropicOCV(SOCInterpolatedHealth):
+class EntropicOCV(SOCInterpolatedHealth, validate_assignment=True):
     base_values: Union[float, np.ndarray] = \
         Field(
             default=0,
             description='Values of entropic OCV term at specified SOCs. Units: V/°C')
 
 
-class OpenCircuitVoltage(HealthVariable):
+class OpenCircuitVoltage(HealthVariable, validate_assignment=True):
```
[the remark this answers is not in the document]
Comment (victorventuri) [reply asoh/models/ecm/components.py:124]: That was my original idea, but I didn't want to change your `HealthVariable` definition, as I don't know if you had any other intended uses for it. I will move this flag to the `HealthVariable` and re-push.

File: asoh/models/ecm/utils.py [hunk 4]
```diff
@@ -20,6 +20,18 @@ class SOCInterpolatedHealth(HealthVariable, validate_assignment=True):
                 'quadratic', 'cubic', 'previous', 'next'] = \
         Field(default='linear', description='Type of interpolation to perform')
 
+    @field_validator('base_values', mode='after')
+    @classmethod
+    def squash_single_input_array(cls, values: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (victorventuri) [reply asoh/models/ecm/utils.py:25]: This in particular was needed for the interpolation, which raises an error if it only has one value to interpolate. But yes, agree.

File: tests/estimators/online/joint/test_ecm_joint.py (hunk 1 above)
Comment (victorventuri) [reply tests/estimators/online/joint/test_ecm_joint.py:48]: I did not either... praise the stackoverflow overlords!

File: asoh/models/ecm/components.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply asoh/models/ecm/components.py]: Just pushed. Marking as resolved.

**Pull request merged**